File: apps/mobile/lib/core/utils/bash_approval_ui.dart
import '../../l10n/generated/app_localizations.dart';

String formatBashApprovalRememberPrefix(String command, {int maxLength = 48}) {
  final trimmed = command.trim();
  if (trimmed.length <= maxLength) {
    return trimmed;
  }
  return '${trimmed.substring(0, maxLength - 1)}…';
}

String buildBashApprovalRememberPrefixLabel(
  String command,
  AppLocalizations l10n,
) {
  return '${l10n.bashApprovalRememberPrefix}'
      '${formatBashApprovalRememberPrefix(command)}';
}

String bashApprovalDenyOptionLabel(AppLocalizations l10n) =>
    l10n.bashApprovalDenyAdjust;

enum BashApprovalChoice { approve, approveRememberPrefix, deny }

String bashApprovalDecisionValue(BashApprovalChoice choice) {
  return switch (choice) {
    BashApprovalChoice.approve => 'approved',
    BashApprovalChoice.approveRememberPrefix => 'approved_remember_prefix',
    BashApprovalChoice.deny => 'denied',
  };
}

File: apps/desktop/src/shared/bash-approval-ui.ts
export type BashApprovalChoice =
  | "approve"
  | "approve_remember_prefix"
  | "deny"
  | "deny_custom";

/**
 * @deprecated Prefer i18n key `approval.bash.otherPlaceholder` in UI.
 * Free-form deny row label (locale-agnostic fallback only).
 */
export const BASH_APPROVAL_DENY_CUSTOM_OPTION_LABEL = "Other";

/** @deprecated Use BASH_APPROVAL_DENY_CUSTOM_OPTION_LABEL */
export const BASH_APPROVAL_DENY_OPTION_LABEL = BASH_APPROVAL_DENY_CUSTOM_OPTION_LABEL;

export function buildBashApprovalChoices(options?: {
  includeRememberPrefix?: boolean;
}): BashApprovalChoice[] {
  const includeRemember = options?.includeRememberPrefix !== false;
  const rows: BashApprovalChoice[] = ["approve"];
  if (includeRemember) {
    rows.push("approve_remember_prefix");
  }
  rows.push("deny", "deny_custom");
  return rows;
}

export function formatBashApprovalRememberPrefix(command: string, maxLength = 48): string {
  const trimmed = command.trim();
  if (trimmed.length <= maxLength) {
    return trimmed;
  }
  return `${trimmed.slice(0, maxLength - 1)}…`;
}

/** Join a localized remember-prefix intro with a truncated command preview. */
export function buildBashApprovalRememberPrefixLabel(
  command: string,
  rememberPrefixIntro: string,
): string {
  return `${rememberPrefixIntro}${formatBashApprovalRememberPrefix(command)}`;
}

export function deriveBashApprovalRememberPrefix(command: string): string {
  return command.trim();
}

export function commandMatchesRememberedBashPrefix(command: string, prefix: string): boolean {
  const trimmedPrefix = prefix.trim();
  if (!trimmedPrefix) {
    return false;
  }
  const trimmedCommand = command.trim();
  return trimmedCommand === trimmedPrefix || trimmedCommand.startsWith(trimmedPrefix);
}

export function commandMatchesAnyRememberedBashPrefix(
  command: string,
  prefixes: readonly string[],
): boolean {
  return prefixes.some((prefix) => commandMatchesRememberedBashPrefix(command, prefix));
}

export function formatBashApprovalDenyMessage(feedback?: string): string {
  const trimmed = feedback?.trim();
  if (!trimmed) {
    return "User denied this Bash command.";
  }
  return `User denied this Bash command. User feedback: ${trimmed}`;
}

export function formatFilesystemApprovalDenyMessage(toolName: string, feedback?: string): string {
  const trimmed = feedback?.trim();
  if (!trimmed) {
    return `User denied this ${toolName} call.`;
  }
  return `User denied this ${toolName} call. User feedback: ${trimmed}`;
}
